src/representation_theory/monoid_algebra_basis.lean

```diff
@@ -0,0 +1,189 @@
+`k[G] ⊗ₖ k[Gⁿ]` (on which `G` acts by `ρ(g₁)(g₂ ⊗ x) = (g₁ * g₂) ⊗ x`) which we will later show
+define an isomorphism. From this we will deduce that `k[Gⁿ⁺¹]` is free over `k[G]`.
+The freeness of these modules means we can use them to build a projective resolution of the
+(trivial) `k[G]`-module `k`, which is useful in group cohomology.
+
+## Main definitions
+
+  * representation.of_mul_action
+  * Rep.of_mul_action
+  * to_tensor
+  * of_tensor
+
+## Implementation notes
+
+We express `k[G]`-module structures on a group `V` as `k`-linear representations of `G` on `V`, as
+this simplifies some proofs and also makes it easier to work with different structures on the same
+`V`. This is because `module` is a class, and there is only one notation for scalar multiplication,
+`•`; representations, meanwhile, give us more flexibility, by avoiding typeclass inference.
+
+We bundle the type `k[Gⁿ]` - or more generally `k[H]` when `H` has `G`-action - and the
+representation together as a term of type `Rep k G`, and we call it `Rep.of_mul_action k G H.` This
+is so we can talk about morphisms of representations.
+
+-/
+
+noncomputable theory
+universes v u
+variables (k G : Type u) [comm_ring k]
+
+open monoid_algebra
+
+namespace representation
+variables [monoid G] (H : Type u) [mul_action G H]
+
+/-- A `G`-action on `H` induces a representation `G →* End(k[H])` in the natural way. -/
+@[simps] def of_mul_action : representation k G (H →₀ k) :=
+{ to_fun := λ g, finsupp.lmap_domain k k ((•) g),
+  map_one' := by { ext x y, dsimp, simp },
+  map_mul' := λ x y, by { ext z w, simp [mul_smul] }}
+
+end representation
+namespace Rep
+section
+variables [monoid G] (H : Type u) [mul_action G H]
+
+/-- Given a `G`-action on `H`, this is `k[H]` bundled with the natural representation
+`G →* End(k[H])` as a term of type `Rep k G.` -/
+abbreviation of_mul_action : Rep k G := Rep.of (representation.of_mul_action k G H)
+
+variables {k G} {n : ℕ}
+
+-- No idea what to call these or where to put them; would move them, but then they can't be private.
+/-- Sends `(g₁, ..., gₙ) ↦ (1, g₁, g₁g₂, ..., g₁...gₙ)` -/
+private def to_prod (f : fin n → G) (i : fin (n + 1)) : G :=
+((list.of_fn f).take i).prod
+
+@[simp] private lemma to_prod_zero (f : fin n → G) :
+  to_prod f 0 = 1 :=
+by simp [to_prod]
+
+private lemma to_prod_succ (f : fin n → G) (j : fin n) :
+  to_prod f j.succ = to_prod f j.cast_succ * (f j) :=
+by simp [to_prod, list.take_succ, list.of_fn_nth_val, dif_pos j.is_lt, ←option.coe_def]
+
+private lemma to_prod_succ' (f : fin (n + 1) → G) (j : fin (n + 1)) :
+  to_prod f j.succ = f 0 * to_prod (fin.tail f) j :=
+by simpa [to_prod]
+
+end
+
+variables [group G] (n : ℕ)
+open_locale tensor_product
+
+/-- The `k`-linear map from `k[Gⁿ⁺¹]` to `k[G] ⊗ₖ k[Gⁿ]` sending `(g₀, ..., gₙ)`
+to `g₀ ⊗ (g₀⁻¹g₁, g₁⁻¹g₂, ..., gₙ₋₁⁻¹gₙ)` -/
+def to_tensor_aux :
+  of_mul_action k G (fin (n + 1) → G) →ₗ[k] monoid_algebra k G ⊗[k] ((fin n → G) →₀ k) :=
```

Comment:
You're mixing `monoid_algebra` and `finsupp` here; is this deliberate?

I would have expected either:
```lean
  of_mul_action k G Gⁿ⁺¹ →ₗ[k] monoid_algebra k G ⊗[k] monoid_algebra k Gⁿ :=
```
or
```lean
  of_mul_action k G Gⁿ⁺¹ →ₗ[k] (G →₀ k) ⊗[k] (Gⁿ →₀ k) :=
```

More generally are there reasons for preferring `_ →₀ k` or `monoid_algebra k _` throughout this file? We should be consistent if we can.

Reply:
I think since finsupp can be used everywhere, it should be; I've removed the remaining `monoid_algebra`s.